test/evp_kdf_test.c

```diff
@@ -18,27 +18,43 @@
 #include <openssl/core_names.h>
 #include "testutil.h"
 
+static EVP_KDF_CTX *get_kdfbyname(const char *name)
+{
+    EVP_KDF *kdf = EVP_KDF_fetch(NULL, name, NULL);
+    EVP_KDF_CTX *kctx = EVP_KDF_CTX_new(kdf);
+
+    EVP_KDF_free(kdf);
+    return kctx;
+}
+
 static int test_kdf_tls1_prf(void)
 {
     int ret;
+    EVP_KDF *kdf;
     EVP_KDF_CTX *kctx = NULL;
-    const EVP_KDF *kdf;
     unsigned char out[16];
+    OSSL_PARAM params[4], *p = params;
     static const unsigned char expected[sizeof(out)] = {
         0x8e, 0x4d, 0x93, 0x25, 0x30, 0xd7, 0x65, 0xa0,
         0xaa, 0xe9, 0x74, 0xc3, 0x04, 0x73, 0x5e, 0xcc
     };
 
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_DIGEST,
+                                            (char *)"sha256", sizeof("sha256"));
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SECRET,
+                                             (unsigned char *)"secret",
+                                             (size_t)6);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SEED,
+                                             (unsigned char *)"seed",
+                                             (size_t)4);
+    *p = OSSL_PARAM_construct_end();
+
     ret =
-        TEST_ptr(kdf = EVP_get_kdfbyname(SN_tls1_prf))
+        TEST_ptr(kdf = EVP_KDF_fetch(NULL, SN_tls1_prf, NULL))
         && TEST_ptr(kctx = EVP_KDF_CTX_new(kdf))
         && TEST_ptr_eq(EVP_KDF_CTX_kdf(kctx), kdf)
         && TEST_str_eq(EVP_KDF_name(kdf), SN_tls1_prf)
-        && TEST_int_gt(EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_SET_MD, EVP_sha256()), 0)
-        && TEST_int_gt(EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_SET_TLS_SECRET,
-                                    "secret", (size_t)6), 0)
-        && TEST_int_gt(EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_ADD_TLS_SEED, "seed",
-                                    (size_t)4), 0)
+        && TEST_true(EVP_KDF_CTX_set_params(kctx, params))
         && TEST_int_gt(EVP_KDF_derive(kctx, out, sizeof(out)), 0)
         && TEST_mem_eq(out, sizeof(out), expected, sizeof(expected));
 
```

Comment:
kdf_free() ?

Reply:
Better is using the utility function created to avoid this issue.